### PR TITLE
CMDCT-3854: Adding mandatory column to 2024 measures table

### PR DESCRIPTION
**Pull request by angelaco11**
### Description
Adding mandatory column with mandatory chip to 2024 measures
Screenshot 2024-07-29 at 1 05 17 PM

CMDCT-3854

---
### How to test
1. Navigate to 2024 coreset and select any adult measure
2. Make sure that the Mandatory column is present and that there are Mandatory chips for the following measures:
```
AMM-AD
CDF-AD
FUA-AD
FUH-AD
FUM-AD
HPCMI-AD
IET-AD
MSC-AD
OUD-AD
SAA-AD
SSD-AD
```

3. Navigate back to 2024 coreset and select child measures. Make sure that every child measure has the mandatory chip
4. Do step 3 for HH measures
5. Make sure that the mandatory column is not present for years 21, 22 and 23

### CODE REVIEW

File: services/ui-src/src/components/Table/columns/measures.tsx [hunk 1]
```diff
@@ -83,6 +84,30 @@ export const measuresColumns = (
         );
       },
     },
+    ...(displayMandatoryColumn
```
Comment (angelaco11) [inline services/ui-src/src/components/Table/columns/measures.tsx:87]: This is a really weird implementation to conditionally render an element in an array, but in my opinion it was the cleanest way to conditionally render this column for 2024 measures. I am fully open to any suggestions if anyone has a better idea!!!

File: services/ui-src/src/components/Table/columns/measures.tsx [hunk 2]
```diff
@@ -55,6 +55,7 @@ const MeasureStatusText = ({
 export const measuresColumns = (
   year: string
 ): TableColumn<MeasureTableItem.Data>[] => {
+  const displayMandatoryColumn = year === "2024";
```
Comment (angelaco11) [inline services/ui-src/src/components/Table/columns/measures.tsx:58]: Is this an acceptable way to render stuff for 2024? I saw this in other parts of the application such as with rendering the `View Combined Rates` button and I couldn't think of a better way to do it.

Comment (BearHanded) [reply services/ui-src/src/components/Table/columns/measures.tsx:58]: this at least needs to be revisited when we hit 2025. I would advise at minimum year >= 2024, and with that said think we need to start tracking some features by year in a centralized place. I think we have a lot of these across the app.

Maybe we centralize a featureByYear file that contains all of the yearly behaviors, and for each function/flag returns true or false based off the provided year?

Comment (angelaco11) [reply services/ui-src/src/components/Table/columns/measures.tsx:58]: Yeah i really like the idea of a `featureByYear` file almost like how we use the commonLabels files for each year in the questionnaires. 

Should we make another ticket for that? Anyone who picks that up could go through and find all the places where we are doing stuff like `year === 2024` and replace it with new `featureByYear` stuff. Would be a bit of an undertaking but worth it in my opinion, especially when we get into 2025 and beyond

Comment (ailZhou) [reply services/ui-src/src/components/Table/columns/measures.tsx:58]: the only other way i can see doing this without using the year is to check whether mandatory even exist in the measure and then filtering that column generation out if it doesn't. 

```
  let columns = QMR.measuresColumns(year);
  if(!measures.some((measure) => measure.mandatory))
    columns = columns.filter((column) => column.header != "Mandatory");
```
But this only works if there's at least one measure that is mandatory in a mandatory coreset 😓

Comment (ailZhou) [reply services/ui-src/src/components/Table/columns/measures.tsx:58]: @angelaco11 just saw your response to Brendan. That ticket would probably end up as being part of the refactoring, so if you make it, just toss it into the QMR Refactor Board.

Comment (angelaco11) [reply services/ui-src/src/components/Table/columns/measures.tsx:58]: Ok yeah I feel like creating that ticket and chuckin on the refactor board sounds most reasonable here. You guys cool with that conclusion?

Comment (angelaco11) [reply services/ui-src/src/components/Table/columns/measures.tsx:58]: here is said ticket for vis https://jiraent.cms.gov/browse/CMDCT-3883

File: services/app-api/types.ts [hunk 3]
```diff
@@ -39,10 +39,12 @@ export interface Measure {
   year: number;
   placeholder?: boolean;
   /**
-   * The `autoCompleted` property is not present on measures in the database;
-   * it is set on fetch, according to the metadata in measureList.ts.
+   * The `autoCompleted` and `mandatory` properties are not present on
+   * measures in the database; they are set on fetch, according to the
+   * metadata in measureList.ts.
```
Comment (benmartin-coforma) [inline services/app-api/types.ts:44]: You added relevant info to my comment, and kept the grammar correct 😭 ❤️ 🙏

File: services/ui-src/src/components/Table/columns/measures.tsx (hunk 1 above)
Comment (benmartin-coforma) [reply services/ui-src/src/components/Table/columns/measures.tsx:87]: I think this is a good choice. It does feel a bit awkward, but the other ideas that popped into my head would be more awkward still 
* Adding it unconditionally but filtering afterward
* Splicing it in afterward, conditionally
* Working in some kind of `shouldDisplay` flag and referencing it later? Somehow?

The advantage of the current approach is that it's 100% inline. Want to know what's in the array? Follow this process:
1. Look where the array is defined
2. You're done.

I appreciate processes like that.

**Review by benmartin-coforma — APPROVED** [review]

File: services/ui-src/src/components/Table/columns/measures.tsx (hunk 1 above)
Comment (angelaco11) [reply services/ui-src/src/components/Table/columns/measures.tsx:87]: Thank you for giving me more ideas on how to do that! The adding it unconditionally but filtering afterward is also a super solid idea that I hadn't considered!! But I agree. with you, in this particular case where we are rendering columns, it is really nice to have it inline even though it's not the most common syntax

File: services/app-api/types.ts (hunk 3 above)
Comment (angelaco11) [reply services/app-api/types.ts:44]: your comment was just so well written that it was that easy to add on to 🫶

**Review by rocio-desantiago — APPROVED** [review]
Very straightforward. Great brainstorming sesh on prospective future years.

**Pull request merged**